### DISCUSSION TITLE
Voodoo 4 and 5 emulation

### DISCUSSION THREAD

**Discussion by Torinde** [86Box/86Box#2909] [Ideas]
At #2710 there is a document [voodoo45_spec.pdf](https://github.com/86Box/86Box/files/9740186/voodoo45_spec.pdf) 

Per [Wikipedia](https://en.wikipedia.org/wiki/3dfx_Interactive#Products) the performance is:
- the already supported 3500SE: 400MTexels/s, 200MPixels/s
- Voodoo 4500: 332MTexels/s **(less than 3500SE)**, 332MPixels/s
- Voodoo 5500: 664MTexels/s, 664MPixels/s

Per [86box docs](https://86box.readthedocs.io/en/latest/settings/display.html#voodoo-graphics) Voodoo emulation is already using multiple threads on the host (although isn't there an error in having less threads for SLI compared to single card?).

Modern top desktop CPUs have 32 threads (without going into Threadripper and server models)

Related: https://github.com/PCBox/PCBox/discussions/89

**Comment by MrPepka** [comment]
I saw this pdf and it mentions only Voodoo, Voodoo2, Banshee and Voodoo3 cards, it is one and two that V4 and V5 emulation would probably be too inefficient considering the power of these cards

**Reply by Torinde** [reply]
The PDF is about "Napalm" (codename for what eventually became VSA-100 (Voodoo4/5).
About emulation performance - numbers above show that Voodoo4 is in the same ballpark as the already supported Voodoo 3500SE/Voodoo 2SLI and Voodoo5 is double of that. Keeping in mind that latest desktop CPUs have much more cores/threads and that 86box Voodoo emulation already uses multi-threading - it seems quite feasible to emulate Voodoo5.

**Reply by Torinde** [reply]
Per [86box docs](https://86box.readthedocs.io/en/latest/settings/display.html#voodoo-graphics)
![image](https://user-images.githubusercontent.com/97228894/205004368-171c6183-90d9-41b6-9089-2e3871ddec78.png)
- Voodoo 2: 180MTexels/s, 90MPixels/s
- Voodoo 2 SLI: 360MTexels/s, 180MPixels/s
- Voodoo 3500SE: 400MTexels/s, 200MPixels/s

If that can be emulated by 4 cores, then Voodoo 5500 should be possible with 8 cores? Also, keep in mind that Voodoo 5500 = 2x Voodoo 4500 physical chips, e.g. workload is definitely splitable.

**Reply by Korkman** [reply]
Yes and no. The actual performance of the currently emulated cards is not as high as the specs as far as I can tell. So doubling the render thread count would get us closer to the true performance of already emulated cards.

**Reply by Torinde** [reply]
Can somebody quantify that? E.g. "current emulation performance is 180MTexels/s and 90MPixels/s when running on 4 Zen2 cores"? @kjliew @Korkman 

Looking at latest desktop CPUs I'm optimistic that sooner rather than later host CPU performance will catch-up.
Voodoo4 4500 and Voodoo5 5500 were released and thus relatively common cards. But the top 3dfx model, four chip Voodoo5 6000 is available only as prototypes - can somebody get hold of ROM from such rare card? @jriwanek already added another rare unreleased model - Velocity 200.

**Reply by Korkman** [reply]
@Torinde The best we can do is probably 3DMark scores (with no less than 100% CPU emulation speed, ofc, because guest CPU emulation can bottleneck and skew the scores) or Quake Demo FPS comparisons with old benchmarks from the real thing. I'm curious myself and will post some comparisons soon.

The Voodoo5 6000 is currently reproduced by Anthony Zxclxiv, btw, and contacting him may provide some intel on their workings (it might be he is only refurbishing chips without dealing with ROMs in any intimate way, but asking helps).

Not saying I wouldn't like these cards added, but the current cards emulated need some serious work done as well. The recompiler is buggy with more than 2 render threads and not properly ported to 64 bit, which is a big headache for Apple Silicon users. A dream come true would be a complete rewrite utilizing CUDA / OpenCL to gain more performance, but it is questionable if that is possible at all.

**Reply by Korkman** [reply]
OK so at least in 3DMark 99 Max, at 800x600, 4 render threads on my Ryzen 5900X deliver performance comparable to a real Voodoo3 3000.

Emulated:
```
Game 1 - Race: 34.8 FPS
Game 2 - First Person: 23.2 FPS
```
Real:
```
Game 1 - Race: 35.7 FPS
Game 2 - First Person: 28.3 FPS
```

Text summary files:
[emulated with 4 threads](https://github.com/86Box/86Box/files/10145354/3dmark.800x600.V3000.PCI.txt)
[reference](https://github.com/86Box/86Box/files/10145413/reference.txt) from [a real Voodoo3 3000 PCI](https://www.vogons.org/viewtopic.php?p=366760#p366760)

3DMark 99 Max benchmark database: [Slot 1 Cyrix 500 V3 3000 PCI.zip](https://github.com/86Box/86Box/files/10145381/Slot.1.Cyrix.500.V3.3000.PCI.zip)

In terms of fill rate (as measured by 3DMark 99 Max), the Voodoo emulation seems to scale well with more threads:

![grafik](https://user-images.githubusercontent.com/3186931/205414819-0c5dcb9b-aebb-4b94-90bd-909272cac613.png)

Unfortunately, not everything scales with render threads. And even when using a 3DNow accelerated CPU (I found the VIA Cyrix III @ 500 MHz to be the fastest I can emulate at full speed) 3DMark 99 seems to be bottlenecked on the emulated CPU.

![grafik](https://user-images.githubusercontent.com/3186931/205418233-7a2570ad-176c-47f9-b201-0d1405ee143b.png)

![grafik](https://user-images.githubusercontent.com/3186931/205418043-36559e74-2e2f-40de-a779-63bc6e393017.png)

So a different perspective on feasibility comes to mind: Will pairing a Voodoo 4 / 5 with a PII 233 MHz or Cyrix III 500 MHz bottleneck on the CPU so hard it won't generate any performance gains?

**Reply by Korkman** [reply]
> Well, look at Tech Tangent's 98 anniversary build on YouTube. It does perform better than that

... with a 1 GHz Pentium III🍒

**Reply by Korkman** [reply]
Here's two excerpts from a benchmark table from [thandor.net](https://thandor.net/object/48)

3DMark2000 800x600
![grafik](https://user-images.githubusercontent.com/3186931/209560601-f2fe74a7-1d42-4f7d-95b8-b6ac202a2268.png)

Note how the results are heavily CPU bound: a PII 450 Mhz, which is double of that which can currently be emulated at 100% speed in 86box, will score basically the same for a Voodoo 2 SLI and a Voodoo 5500.

86box scores 1362 with my 450 MHz VIA Cyrix III (a CPU vastly inferior to a PII at 450 MHz! It is only the 3DNow! support helping the score here!) and Voodoo Banshee.

Compare that to the 1.4 GHz PIII series:

![grafik](https://user-images.githubusercontent.com/3186931/209560661-b0fb8898-68be-4b8c-a302-977fa9c10580.png)

So it will be a loooong time until we would see performance gains from emulating V5 _in terms of 3DMark scores_. I'll see about possible scaling in games next.

**Reply by Korkman** [reply]
Conveniently, the same website also lists game benchmarks. Here's the [Voodoo5 5500](https://thandor.net/object/17) review, which I distilled these PII 450 results from:
![grafik](https://user-images.githubusercontent.com/3186931/209571218-3c43a735-a65e-4e6e-aaef-5e8bf30c550f.png)

![grafik](https://user-images.githubusercontent.com/3186931/209571280-d8a56716-b7b1-4210-a592-cc34fba45c4d.png)

![grafik](https://user-images.githubusercontent.com/3186931/209571347-b5afa567-43a4-466a-8f58-8f6d6b44dbba.png)

![grafik](https://user-images.githubusercontent.com/3186931/209571408-2ac8756c-f425-4f6a-bd0a-3da9c7d5a0fe.png)

Interesting. Most games are not as CPU bound as 3DMark and as such there are some performance uplifts to be seen. Weirdly enough, Unreal throws too much at the CPU when paired with V4 or V5, causing the frame rate to actually drop.

I think this answers the question whether **_physical_** V4 or V5 cards would be beneficial when paired to a CPU of the era emulated by 86box. Yes, to some extend, games would gain performance.

**Reply by Torinde** [reply]
Great. 
Also it seems top line modern multi-core CPU has enough power to emulate any Voodoo. 
On the other hand modern CPU host core is not yet fast enough to emulating the respective guest CPU needed to use such Voodoo (above 1GHz) - unless resorting to a hypervisor core, for example based on [virt86](https://github.com/StrikerX3/virt86).

**Reply by Korkman** [reply]
> Also it seems top line modern multi-core CPU has enough power to emulate any Voodoo

It depends. Here's some final input on this: I've benchmarked Dethkarz, a game which I think is not bottlenecked on the emulated CPU, with similar settings to thandor's, targeting 30 fps instead of 60 to get results more quickly:
```
Detail Level: Extreme
Resolution: 1024x768 16-bit
Target Frame Rate: 30 FPS
Allow Slow Down: None

22 fps @ 1 rt, target 30fps
33 fps @ 2 rt, target 30fps = 50% increase
42 fps @ 4 rt, target 30fps = 27% increase
40 fps @ 4 rt, target 60fps (to show the fps target doesn't limit the fps)
```
At 4 render threads, the performance is approaching that of a real Voodoo3 1000 (44 fps), still way behind that of a Voodoo3 3500TV (61 fps). And I'm afraid the current code doesn't scale in a linear fashion - there's a bottleneck somewhere, creating a single-threaded problem. Simply cranking up the thread count will not increase performance.

AFAIK, the way Voodoo graphics accelerators are emulated allows for arbitrary performance for any model, until artificially limiting performance becomes necessary. So basically, for better performance, there is no need for a Voodoo 4/5 emulation. What's required is a stern look at the Voodoo emulation core, which may already have hit a dead end where a single-threaded problem cannot be distributed any more, similar to the situation of CPU emulation.

My conclusion is that some games like Dethkarz will indeed profit from faster Voodoo emulation. Whether this is possible, only devs can tell.

**Reply by Torinde** [reply]
It's unlikely, but if somebody gets hold of [Larrabee](https://en.wikipedia.org/wiki/Larrabee_(microarchitecture)) drivers - maybe that will help as they are basically a 32 core x86-to-DirectX/OpenGL converter... It seems people got hold of prototype cards and are even trying to decompile its BIOS: [1](https://community.intel.com/t5/Graphics/About-intel-Larrabee-graphics-cards/td-p/1241910), [2](https://www.tomshardware.com/news/working-intel-larrabee-graphics-card-found-and-filmed), [3](https://www.vogons.org/viewtopic.php?t=57207)

**Reply by Cacodemon345** [reply]
That is ridiculously out of scope for our emulator.

**Reply by WhiteMagicRaven** [reply]
https://bitsavers.trailing-edge.com/components/3dfx/

Voodoo4
Voodoo5
Documentations
